snowdrop-execution/src/Snowdrop/Execution/DbActions/AVLp/Accum.hs

```diff
@@ -45,111 +52,153 @@ data AVLChgAccum' h k v = AVLChgAccum
-modAccum' (AVLChgAccum initAvl initAcc initTouched) cMod pState = do
-    returnWithUndo =<< case cMod of
-      CAMChange (changeSetToList -> cs) ->
-        ( Right . Just . doUncurry AVLChgAccum
+    => ctx
+    -> AVLChgAccum h t
+    -> ChgAccumModifier '[t]
+    -> m (Either CSMappendException (AVLChgAccum h t, Undo '[t]))
+modAccumOne pState (AVLChgAccum initAvl initAcc initTouched) cMod = reThrowAVLEx @(HKey t) $
+  returnWithUndo =<< case cMod of
+    CAMChange (hChangeSetElToList . unone -> cs) ->
+        ( Right . doUncurry AVLChgAccum
           <$> runAVLCacheT (foldM modAVL (initAvl, initTouched) cs) initAcc pState )
             `catch` \e@(CSMappendException _) -> pure $ Left e
-      CAMRevert (Undo _cs sn) ->
-        case (BS.null sn, deserialise sn) of
-          (False, Left str) ->
+    CAMRevert (Undo _cs (unAvlRevision . unone -> sn)) -> case (BS.null sn, deserialise sn) of
+        (False, Left str) ->
             throwM $ DbApplyException $ "Error parsing AVL snapshot from undo: " <> toText str
-          (False, Right rootH) ->
-            pure $ Right $ Just $ AVLChgAccum (mkAVL rootH) initAcc initTouched
-          _ -> pure $ Right $ Just $ AVLChgAccum initAvl initAcc initTouched
+        (False, Right rootH) ->
+            pure $ Right $ AVLChgAccum (mkAVL rootH) initAcc initTouched
+        _ -> pure $ Right $ AVLChgAccum initAvl initAcc initTouched
   where
     returnWithUndo
-        :: Either (CSMappendException k) (AVLChgAccum h k v)
-        -> m (Either (CSMappendException k) (AVLChgAccum h k v, Undo k v))
+        :: Either CSMappendException (AVLChgAccum h t)
+        -> m (Either CSMappendException (AVLChgAccum h t, Undo '[t]))
     returnWithUndo (Left e)   = pure $ Left e
     returnWithUndo (Right cA) =
-        pure $ Right (cA, Undo def $ serialise $ avlRootHash initAvl) -- TODO compute undo
-
+        pure $ Right (cA, Undo def $ rone $ AvlRevision $ serialise $ avlRootHash initAvl) -- TODO compute undo
 
     doUncurry :: (a -> b -> c -> d) -> ((a, c), b) -> d
     doUncurry f ((a, c), b) = f a b c
 
     modAVL
-        :: (AVL.Map h k v, Set h)
+        :: (KVConstraint k v, Serialisable (MapLayer h k v h), AVL.Hash h k v)
+        => (AVL.Map h k v, Set h)
         -> (k, ValueOp v)
         -> AVLCacheT h (ReaderT ctx m) (AVL.Map h k v, Set h)
     modAVL (avl, touched) (k, valueop) = processResp =<< AVL.lookup' k avl
       where
         processResp ((lookupRes, (<> touched) -> touched'), avl') =
           case (valueop, lookupRes) of
-            (NotExisted ,_     ) -> pure (avl', touched')
-            (New v      ,_     ) -> (, touched') . snd <$> AVL.insert' k v avl'
-            (Rem        ,Just _) -> (, touched') . snd <$> AVL.delete' k avl'
-            (Upd v      ,Just _) -> (, touched') . snd <$> AVL.insert' k v avl'
+            (NotExisted, _     ) -> pure (avl', touched')
+            (New v     , _     ) -> (, touched') . snd <$> AVL.insert' k v avl'
```

Comment:
Isn't is a bug? `New` ignores if value existed before.

Reply:
looks so, cc @georgeee

Reply:
Agree, there shall be two `Nothing `

Reply:
@georgeee and for `NotExisted` too?

Reply:
Right